subprojects/base-services/src/main/java/org/gradle/internal/FileUtils.java

```diff
@@ -16,20 +16,41 @@
 
 package org.gradle.internal;
 
+import com.google.common.collect.Lists;
 import org.apache.commons.io.FilenameUtils;
+import org.apache.commons.lang.StringUtils;
 import org.gradle.api.GradleException;
 import org.gradle.api.UncheckedIOException;
 
 import java.io.File;
 import java.io.IOException;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.Comparator;
 import java.util.Iterator;
-import java.util.LinkedList;
 import java.util.List;
 
 public class FileUtils {
     public static final int WINDOWS_PATH_LIMIT = 260;
 
+    private static final Comparator<File> FILE_SEGMENT_COMPARATOR = new Comparator<File>() {
+        @Override
+        public int compare(File left, File right) {
```

Comment:
Perhaps we can consider using here a tweaked version of the implementation of `compareTo` in `String` which treats `File.separatorChar` as the smallest character. Its algorithm is similar in shape to this one (comparing characters instead of segments) and is of similar length. Therefore, I don't think it will be confusing, and it has the benefit of allocating fewer short-lived objects.